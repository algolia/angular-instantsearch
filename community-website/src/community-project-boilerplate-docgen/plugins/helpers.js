const highlight = require('./syntaxHighlighting.js');
const md = require('./mdRenderer.js');

// this plugin provides ATM one helper to easily compute the publicPath of assets
module.exports = function helpers(filenames, metalsmith, cb) {
  // eslint-disable-next-line no-param-reassign
  metalsmith.metadata().h = {
    markdown(src) {
      return md.render(src);
    },
    highlight(src, opts) {
      const lang = opts && opts.lang;
      return highlight(src, lang);
    },
    maybeActive(navPath, singlePathOrArrayOfPaths) {
      const pathsToTest = [].concat(singlePathOrArrayOfPaths);
      return pathsToTest.some(pathToTest => navPath.indexOf(pathToTest) === 0)
        ? 'active'
        : '';
    },
  };

  cb();
};
