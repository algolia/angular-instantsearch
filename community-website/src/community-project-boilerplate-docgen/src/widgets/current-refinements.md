---
title: Widgets - CurrentRefinements
layout: widget.pug
---

# Current Refinements

## Description

The current refinements widget has two purposes:

* give the user a synthetic view of the current filters.
* give the user the ability to remove a filter.

This widget is usually in the top part of the search UI.

## Options

| Attributes       | Type                            | Description
| -                | -                               | -
| `attributes?`    | `{name: string, label: string}` | Label definitions for the different filters
| `onlyListedAttributes?` | `boolean` | Only use the declared attributes. By default, the widget displays the refinements for the whole search state. If true, the list of `attributes` in attributes is used
| `clearRefinements?`      | `"before" / "after" / boolean`  | Defines the clear all button position. By default, it is placed before the set of current filters. If the value is false, the button wonâ€™t be added in the widget
| `clearsQuery?`   | `boolean`                       | If true, the clear all button also clears the active search query.
| `clearRefinementsLabel?` | `string`                        | Label used on the clear all button.
| `autoHideContainer?` | `boolean`  | Hides the widget if there's no refinements to display

## Code example

```html
<ais-current-refinements [clearsQuery]="true">
</ais-current-refinements>
```
