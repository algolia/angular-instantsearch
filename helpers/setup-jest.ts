import "jest-preset-angular";

// stupid workaround for TS
import * as helper from "algoliasearch-helper";
require("algoliasearch-helper").default = helper;

const mock = () => {
  let storage = {};
  return {
    getItem(key) {
      return storage[key] || null;
    },
    setItem(key, value) {
      storage[key] = value;
    },
    removeItem(key) {
      delete storage[key];
    },
    clear() {
      storage = {};
    }
  };
};

Object.defineProperty(window, "localStorage", { value: mock() });
Object.defineProperty(window, "sessionStorage", { value: mock() });
Object.defineProperty(window, "getComputedStyle", {
  value: () => ["-webkit-appearance"]
});
