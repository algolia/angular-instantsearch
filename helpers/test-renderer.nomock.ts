import { Component, ViewChild } from '@angular/core';
import { TestBed } from '@angular/core/testing';

import { NgAisInstantSearchModule } from '../src/instantsearch/instantsearch.module';

// same as test-renderer, but without mocking InstantSearch
// In the future, we want to migrate all tests to this

export function createRenderer({
  template,
  TestedWidget,
  defaultState,
  additionalImports,
  additionalDeclarations,
  methods = {},
}: {
  template: string;
  TestedWidget: any;
  additionalImports?: any[];
  additionalDeclarations?: any[];
  defaultState?: {};
  methods?: {};
}) {
  return function(state?: {}, isFirstRender = false) {
    return render(
      {
        template,
        TestedWidget,
        additionalImports,
        additionalDeclarations,
        state: state ? { ...(defaultState || {}), ...state } : undefined,
        methods,
      },
      isFirstRender
    );
  };
}

function render(
  {
    template,
    TestedWidget,
    additionalImports,
    additionalDeclarations,
    state,
    methods,
  }: {
    template: string;
    TestedWidget: any;
    additionalImports?: any[];
    additionalDeclarations?: any[];
    state?: {};
    methods?: {};
  },
  isFirstRender = false
) {
  @Component({
    template: `
      <ais-instantsearch [config]="config">
        ${template}
      </ais-instantsearch>
    `,
  })
  class TestContainer {
    config = {
      searchClient: {
        search(requests) {
          return Promise.resolve({
            results: requests.map(() => ({ hits: [] })),
          });
        },
      },
      indexName: 'instant_search',
    };
    @ViewChild(TestedWidget) testedWidget;
    constructor() {
      Object.keys(methods).forEach(methodName => {
        this[methodName] = methods[methodName];
      });
    }
  }

  TestBed.configureCompiler({
    preserveWhitespaces: false,
  } as any).configureTestingModule({
    declarations: [
      TestContainer,
      TestedWidget,
      ...(additionalDeclarations || []),
    ],
    imports: [NgAisInstantSearchModule.forRoot(), ...(additionalImports || [])],
  });

  const fixture = TestBed.createComponent(TestContainer);
  fixture.detectChanges();

  if (state) {
    fixture.componentInstance.testedWidget.updateState(state, isFirstRender);
    fixture.detectChanges();
  }

  return fixture;
}
