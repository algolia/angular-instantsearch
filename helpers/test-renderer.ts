import { Component, ViewChild } from "@angular/core";
import { TestBed } from "@angular/core/testing";

import { NgAisInstantSearchModule } from "../src/instantsearch/instantsearch.module";

// mock component, dont create a real instantearch instance
jest.mock("../src/base-widget");
jest.mock("../src/instantsearch/instantsearch");

export function createRenderer({
  template,
  TestedWidget,
  defaultState,
  additionalImports,
  additionalDeclarations
}: {
  template: string;
  TestedWidget: any;
  additionalImports?: any[];
  additionalDeclarations?: any[];
  defaultState?: {};
}) {
  return function(state?: {}, firstRender = false) {
    return render(
      {
        template,
        TestedWidget,
        additionalImports,
        additionalDeclarations,
        state: state ? { ...(defaultState || {}), ...state } : undefined
      },
      firstRender
    );
  };
}

function render(
  {
    template,
    TestedWidget,
    additionalImports,
    additionalDeclarations,
    state
  }: {
    template: string;
    TestedWidget: any;
    additionalImports?: any[];
    additionalDeclarations?: any[];
    state?: {};
  },
  firstRender = false
) {
  @Component({
    template: `
      <ng-ais-instantsearch>
        ${template}
      </ng-ais-instantsearch>
    `
  })
  class TestContainer {
    @ViewChild(TestedWidget) testedWidget;
  }

  TestBed.configureTestingModule({
    declarations: [
      TestContainer,
      TestedWidget,
      ...(additionalDeclarations || [])
    ],
    imports: [NgAisInstantSearchModule.forRoot(), ...(additionalImports || [])]
  });

  const fixture = TestBed.createComponent(TestContainer);
  fixture.detectChanges();

  if (state) {
    fixture.componentInstance.testedWidget.updateState(state, firstRender);
    fixture.detectChanges();
  }

  return fixture;
}
