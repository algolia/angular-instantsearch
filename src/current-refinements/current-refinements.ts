import { Component, Input, Inject, forwardRef } from "@angular/core";

import { connectCurrentRefinedValues } from "instantsearch.js/es/connectors";
import { noop, isFunction } from "lodash-es";

import { BaseWidget } from "../base-widget";
import { NgAisInstance } from "../instantsearch/instantsearch-instance";
import { NgAisInstantSearch } from "../instantsearch/instantsearch";

export type CurrentRefinementsState = {
  attributes: {};
  clearAllClick: Function;
  clearAllURL: Function;
  createURL: Function;
  refine: Function;
  refinements: {}[];
};

@Component({
  selector: "ng-ais-current-refinements",
  template: `
    <div
      [class]="cx()"
      *ngIf="!isHidden"
    >
      <button
        [class]="cx('reset')"
        (click)="handleClearAllClick($event)"
        *ngIf="clearRefinements === 'before' || clearRefinements === true">
        {{clearRefinementsLabel}}
      </button>

      <ul [class]="cx('list')">
        <li
          [class]="cx('item')"
          *ngFor="let refinement of refinements"
          (click)="handleClick($event, refinement)"
        >
          <button [class]="cx('button')">
            <span [class]="cx('label')">{{refinement.computedLabel}}</span>
            <span
              *ngIf="refinement.count && refinement.count > 0"
              [class]="cx('count')"
            >
              {{refinement.count}}
            </span>
            <span [class]="cx('delete')">âœ•</span>
          </button>
        </li>
      </ul>

      <button
        [class]="cx('reset')"
        (click)="handleClearAllClick($event)"
        *ngIf="clearRefinements === 'after'">
        {{clearRefinementsLabel}}
      </button>
    </div>
  `
})
export class NgAisCurrentRefinements extends BaseWidget {
  // render options
  @Input() public clearRefinements: "before" | "after" | boolean = "after";
  @Input() public clearRefinementsLabel: string = "Clear refinements";
  @Input() public transformItems?: Function;

  // connector options
  @Input() public onlyListedAttributes: boolean = false;
  @Input() public clearsQuery: boolean = false;
  @Input()
  public attributes: {
    name: string;
    label: string;
  }[] = [];

  public state: CurrentRefinementsState = {
    attributes: {},
    clearAllClick: noop,
    clearAllURL: noop,
    createURL: noop,
    refine: noop,
    refinements: []
  };

  get isHidden() {
    return this.state.refinements.length === 0 && this.autoHideContainer;
  }

  get refinements() {
    return isFunction(this.transformItems)
      ? this.transformItems(this.state.refinements)
      : this.state.refinements;
  }

  constructor(
    @Inject(forwardRef(() => NgAisInstantSearch))
    public instantSearchParent: any
  ) {
    super("CurrentRefinements");
  }

  public ngOnInit() {
    this.createWidget(connectCurrentRefinedValues, {
      attributes: this.attributes,
      clearsQuery: this.clearsQuery,
      onlyListedAttributes: this.onlyListedAttributes
    });
    super.ngOnInit();
  }

  public handleClick(event: MouseEvent, refinement: {}) {
    event.preventDefault();
    this.state.refine(refinement);
  }

  public handleClearAllClick(event: MouseEvent) {
    event.preventDefault();
    this.state.clearAllClick();
  }
}
