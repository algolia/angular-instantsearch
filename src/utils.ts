export function bem(widgetName: string) {
  const cx = function(element?: string, subElement?: string) {
    if (element) {
      const scoppedWidgetName = `ais-${widgetName}-${element}`;

      // output `ais-Widget-Header|Body|Footer ais-Header|Body|Footer`
      if (element === "header" || element === "body" || element === "footer") {
        const nonScoppedWidgetName = `ais-${element}`;
        return `${scoppedWidgetName} ${nonScoppedWidgetName}`;
      }

      // output `ais-Widget-Xyz--abc`
      if (subElement) {
        return `${scoppedWidgetName}--${subElement}`;
      }

      // output `ais-Widget-Xyz`
      return scoppedWidgetName;
    }

    // output `ais-Widget`
    return `ais-${widgetName}`;
  };
  return cx;
}

export function parseNumberInput(input?: number | string) {
  return typeof input === "string" ? parseInt(input, 10) : input;
}
