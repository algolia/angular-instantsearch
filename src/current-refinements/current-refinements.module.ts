import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';

import { NgAisCurrentRefinements } from './current-refinements';

@NgModule({
  declarations: [NgAisCurrentRefinements],
  entryComponents: [NgAisCurrentRefinements],
  exports: [NgAisCurrentRefinements],
  imports: [CommonModule],
})
export class NgAisCurrentRefinementsModule {}
