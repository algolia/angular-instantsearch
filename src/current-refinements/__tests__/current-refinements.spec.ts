import { createRenderer } from '../../../helpers/test-renderer';
import { NgAisCurrentRefinements } from '../current-refinements';

import { bem } from '../../utils';

const cx = bem('CurrentRefinements');

const defaultState = {
  attributes: {},
  clearAllClick: jest.fn(),
  clearAllURL: jest.fn(),
  createURL: jest.fn(),
  refined: jest.fn(),
  refinements: [
    {
      type: 'disjunctive',
      attributeName: 'brand',
      name: 'Canon',
      count: 27,
      exhaustive: true,
      computedLabel: 'Canon',
    },
    {
      type: 'disjunctive',
      attributeName: 'brand',
      name: 'Sony',
      count: 28,
      exhaustive: true,
      computedLabel: 'Sony',
    },
    {
      type: 'hierarchical',
      attributeName: 'hierarchicalCategories.lvl0',
      name: 'Cameras & Camcorders',
      count: 55,
      computedLabel: 'Cameras & Camcorders',
    },
    {
      type: 'numeric',
      attributeName: 'popularity',
      name: '0',
      numericValue: 0,
      operator: '>=',
      computedLabel: 'â‰¥ 0',
    },
  ],
};

const render = createRenderer({
  defaultState,
  template: '<ais-current-refinements></ais-current-refinements>',
  TestedWidget: NgAisCurrentRefinements,
});

describe('CurrentRefinedValues', () => {
  it('renders markup without state', () => {
    const fixture = render();
    expect(fixture).toMatchSnapshot();
  });

  it('renders markup with state', () => {
    const fixture = render({});
    expect(fixture).toMatchSnapshot();
  });

  it('should call refine() when clicking on an refined element', () => {
    const refine = jest.fn();
    const fixture = render({ refine });

    const [firstEl] = fixture.debugElement.nativeElement.querySelectorAll(
      'span > button'
    );
    firstEl.click();

    expect(refine).toHaveBeenCalled();
    expect(refine).toHaveBeenCalledWith(defaultState.refinements[0]);
  });

  it('should call clearAllClick() when clicking on clear all', () => {
    const clearAllClick = jest.fn();
    const fixture = render({ clearAllClick });

    fixture.debugElement.nativeElement.querySelector(`.${cx('reset')}`).click();

    expect(clearAllClick).toHaveBeenCalled();
  });

  it('should apply `transformItems` if specified', () => {
    const fixture = render({});

    fixture.componentInstance.testedWidget.transformItems = items =>
      items.map(item => ({
        ...item,
        computedLabel: `foo - ${item.computedLabel}`,
      }));
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should be hidden with `autoHideContainer`', () => {
    const fixture = render({ refinements: [] });
    fixture.componentInstance.testedWidget.autoHideContainer = true;
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });
});
