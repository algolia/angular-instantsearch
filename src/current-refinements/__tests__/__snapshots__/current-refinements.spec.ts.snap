// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrentRefinedValues renders markup with state 1`] = `
<ng-component
  testedWidget={[Function NgAisCurrentRefinements]}
>
  
        <ng-ais-instantsearch>
    
          
    <ng-ais-current-refinements>
      
      
      
      <div
        class="ais-CurrentRefinements"
      >
        
        
        
        
  
        
        
        <ul
          class="ais-CurrentRefinements-list"
        >
          
          
          <li
            class="ais-CurrentRefinements-item"
          >
            
            
            <span
              class="ais-CurrentRefinements-label"
            >
              Brand:
            </span>
            
  
            
            
            <span
              class="ais-CurrentRefinements-category"
            >
              
              
              <span
                class="ais-CurrentRefinements-categoryLabel"
              >
                Canon
              </span>
              
              
              <button
                class="ais-CurrentRefinements-delete"
              >
                ✕
              </button>
              
            
            </span>
            <span
              class="ais-CurrentRefinements-category"
            >
              
              
              <span
                class="ais-CurrentRefinements-categoryLabel"
              >
                Sony
              </span>
              
              
              <button
                class="ais-CurrentRefinements-delete"
              >
                ✕
              </button>
              
            
            </span>
            
          
          </li>
          
        
        </ul>
        <ul
          class="ais-CurrentRefinements-list"
        >
          
          
          <li
            class="ais-CurrentRefinements-item"
          >
            
            
            <span
              class="ais-CurrentRefinements-label"
            >
              Hierarchicalcategories.lvl0:
            </span>
            
  
            
            
            <span
              class="ais-CurrentRefinements-category"
            >
              
              
              <span
                class="ais-CurrentRefinements-categoryLabel"
              >
                Cameras & Camcorders
              </span>
              
              
              <button
                class="ais-CurrentRefinements-delete"
              >
                ✕
              </button>
              
            
            </span>
            
          
          </li>
          
        
        </ul>
        <ul
          class="ais-CurrentRefinements-list"
        >
          
          
          <li
            class="ais-CurrentRefinements-item"
          >
            
            
            <span
              class="ais-CurrentRefinements-label"
            >
              Popularity:
            </span>
            
  
            
            
            <span
              class="ais-CurrentRefinements-category"
            >
              
              
              <span
                class="ais-CurrentRefinements-categoryLabel"
              >
                0
              </span>
              
              
              <button
                class="ais-CurrentRefinements-delete"
              >
                ✕
              </button>
              
            
            </span>
            
          
          </li>
          
        
        </ul>
        
  
        
        
        <button
          class="ais-CurrentRefinements-reset"
        >
          
          Clear refinements
        
        </button>
        
      
      </div>
      
    
    </ng-ais-current-refinements>
    
        
  </ng-ais-instantsearch>
      
</ng-component>
`;

exports[`CurrentRefinedValues renders markup without state 1`] = `
<ng-component
  testedWidget={[Function NgAisCurrentRefinements]}
>
  
        <ng-ais-instantsearch>
    
          
    <ng-ais-current-refinements>
      
      
      
      <div
        class="ais-CurrentRefinements"
      >
        
        
        
        
  
        
        
        
  
        
        
        <button
          class="ais-CurrentRefinements-reset"
        >
          
          Clear refinements
        
        </button>
        
      
      </div>
      
    
    </ng-ais-current-refinements>
    
        
  </ng-ais-instantsearch>
      
</ng-component>
`;

exports[`CurrentRefinedValues should apply \`transformItems\` if specified 1`] = `
<ng-component
  testedWidget={[Function NgAisCurrentRefinements]}
>
  
        <ng-ais-instantsearch>
    
          
    <ng-ais-current-refinements>
      
      
      
      <div
        class="ais-CurrentRefinements"
      >
        
        
        
        
  
        
        
        <ul
          class="ais-CurrentRefinements-list"
        >
          
          
          <li
            class="ais-CurrentRefinements-item"
          >
            
            
            <span
              class="ais-CurrentRefinements-label"
            >
              Brand:
            </span>
            
  
            
            
            <span
              class="ais-CurrentRefinements-category"
            >
              
              
              <span
                class="ais-CurrentRefinements-categoryLabel"
              >
                Canon
              </span>
              
              
              <button
                class="ais-CurrentRefinements-delete"
              >
                ✕
              </button>
              
            
            </span>
            <span
              class="ais-CurrentRefinements-category"
            >
              
              
              <span
                class="ais-CurrentRefinements-categoryLabel"
              >
                Sony
              </span>
              
              
              <button
                class="ais-CurrentRefinements-delete"
              >
                ✕
              </button>
              
            
            </span>
            
          
          </li>
          
        
        </ul>
        <ul
          class="ais-CurrentRefinements-list"
        >
          
          
          <li
            class="ais-CurrentRefinements-item"
          >
            
            
            <span
              class="ais-CurrentRefinements-label"
            >
              Hierarchicalcategories.lvl0:
            </span>
            
  
            
            
            <span
              class="ais-CurrentRefinements-category"
            >
              
              
              <span
                class="ais-CurrentRefinements-categoryLabel"
              >
                Cameras & Camcorders
              </span>
              
              
              <button
                class="ais-CurrentRefinements-delete"
              >
                ✕
              </button>
              
            
            </span>
            
          
          </li>
          
        
        </ul>
        <ul
          class="ais-CurrentRefinements-list"
        >
          
          
          <li
            class="ais-CurrentRefinements-item"
          >
            
            
            <span
              class="ais-CurrentRefinements-label"
            >
              Popularity:
            </span>
            
  
            
            
            <span
              class="ais-CurrentRefinements-category"
            >
              
              
              <span
                class="ais-CurrentRefinements-categoryLabel"
              >
                0
              </span>
              
              
              <button
                class="ais-CurrentRefinements-delete"
              >
                ✕
              </button>
              
            
            </span>
            
          
          </li>
          
        
        </ul>
        
  
        
        
        <button
          class="ais-CurrentRefinements-reset"
        >
          
          Clear refinements
        
        </button>
        
      
      </div>
      
    
    </ng-ais-current-refinements>
    
        
  </ng-ais-instantsearch>
      
</ng-component>
`;

exports[`CurrentRefinedValues should be hidden with \`autoHideContainer\` 1`] = `
<ng-component
  testedWidget={[Function NgAisCurrentRefinements]}
>
  
        <ng-ais-instantsearch>
    
          
    <ng-ais-current-refinements>
      
      
      
      
    
    </ng-ais-current-refinements>
    
        
  </ng-ais-instantsearch>
      
</ng-component>
`;
